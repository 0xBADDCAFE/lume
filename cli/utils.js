import { existsSync } from "../deps/fs.js";
import lume from "../mod.js";
import { join, resolve } from "../deps/path.js";
import { brightGreen } from "../deps/colors.js";

/**
 * @return {Promise<*>} a lume instance - ready to build, run, etc.
 */
export async function buildSite(options, site) {
  options.root = resolve(Deno.cwd(), options.root);
  const configFile = join(options.root, options.config);

  if (!site) {
    if (existsSync(configFile)) {
      const mod = await import(`file://${configFile}`);
      site = mod.default;
    } else {
      site = lume();
    }
  }

  site.options.cwd = options.root;

  if (options.dev) {
    site.options.dev = options.dev;
  }

  if (options.location) {
    site.options.location = new URL(options.location);
  }

  if (options.src) {
    site.options.src = options.src;
  }

  if (options.dest) {
    site.options.dest = options.dest;
  }

  if (options["--"]) {
    site.options.flags = options["--"];
  }

  return site;
}

/**
 * @param command name of the command you're validating
 * @param args array of cli arguments (i.e. options._)
 * @param max the max number of args
 * @param min the min number of args
 */
export function validateArgsCount(command, args, max, min = 0) {
  if (args.length > max) {
    throw new Error(`Unexpected arguments: ${args.slice(max).join(", ")}
    
    Run ${brightGreen(`lume ${command} --help`)} for usage information
    `);
  }

  if (args.length < min) {
    throw new Error(`Missing arguments

    Run ${brightGreen(`lume ${command} --help`)} for usage information
    `);
  }
}
