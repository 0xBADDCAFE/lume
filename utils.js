import { bold, red } from "./deps/colors.js";
import { SEP } from "./deps/path.js";

export async function concurrent(iterable, iteratorFn, limit = 200) {
  const executing = [];

  for await (const item of iterable) {
    const p = iteratorFn(item).then(() =>
      executing.splice(executing.indexOf(p), 1)
    );

    executing.push(p);

    if (executing.length >= limit) {
      await Promise.race(executing);
    }
  }

  await Promise.all(executing);
}

export const mimes = new Map([
  [".html", "text/html; charset=utf-8"],
  [".txt", "text/plain; charset=utf-8"],
  [".js", "text/javascript; charset=utf-8"],
  [".css", "text/css; charset=utf-8"],
  [".atom", "application/atom+xml; charset=utf-8"],
  [".rss", "application/rss+xml; charset=utf-8"],
  [".svg", "image/svg+xml; charset=utf-8"],
  [".json", "application/json"],
  [".webmanifest", "application/manifest+json"],
  [".ico", "image/x-icon"],
  [".avif", "image/avif"],
  [".png", "image/png"],
  [".apng", "image/apng"],
  [".jpg", "image/jpg"],
  [".jpeg", "image/jpg"],
  [".gif", "image/gif"],
  [".mp3", "audio/mpeg"],
  [".mp4", "video/mpeg"],
  [".xml", "text/xml"],
  [".woff", "font/woff"],
  [".woff2", "font/woff2"],
  [".wasm", "application/wasm"],
  [".webp", "image/webp"],
  [".webm", "video/webm"],
  [".zip", "application/zip"],
]);

export function error(context, message, exception) {
  console.error(bold(red(`${context}:`)), message);
  if (exception) {
    console.error(exception);
  }
  console.error("");
}

export function merge(defaults, user) {
  const merged = { ...defaults };

  if (!user) {
    return merged;
  }

  for (const [key, value] of Object.entries(user)) {
    if (
      typeof merged[key] === "object" && typeof value === "object" &&
      !Array.isArray(merged[key])
    ) {
      merged[key] = merge(merged[key], value);
      continue;
    }

    merged[key] = value;
  }

  return merged;
}

export function normalizePath(path) {
  // Is Windows path
  if (SEP !== "/") {
    return path.replaceAll(SEP, "/");
  }

  return path;
}

export const slugChars = new Map([
  ["รฐ", "d"],
  ["ร", "ss"],
]);

export function slugify(string) {
  return string.toLowerCase()
    .normalize("NFKD")
    .replaceAll(/[\s_-]+/g, "-")
    .replaceAll(/[^\w\/-]/g, (char) => slugChars.get(char) || "");
}

export function searchByExtension(path, extensions) {
  for (const [key, value] of extensions) {
    if (path.endsWith(key)) {
      return [key, value];
    }
  }
}
