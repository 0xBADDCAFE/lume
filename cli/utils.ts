import lume from "../mod.ts";
import { toFileUrl } from "../deps/path.ts";
import { dim, red } from "../deps/colors.ts";
import { getConfigFile } from "../core/utils.ts";

import type { Site } from "../core.ts";

/** Create a site instance */
export async function createSite(root: string, config?: string): Promise<Site> {
  const path = await getConfigFile(root, config);

  if (path) {
    console.log(`Loading config file ${dim(path)}`);
    console.log();
    const mod = await import(toFileUrl(path).href);
    return mod.default;
  }

  return lume();
}

/** A list of the available optional plugins */
export const pluginNames = [
  "attributes",
  "base_path",
  "code_highlight",
  "date",
  "esbuild",
  "eta",
  "imagick",
  "inline",
  "jsx",
  "lightningcss",
  "liquid",
  "metas",
  "minify_html",
  "modify_urls",
  "multilanguage",
  "nano_jsx",
  "netlify_cms",
  "on_demand",
  "postcss",
  "prism",
  "pug",
  "relations",
  "relative_urls",
  "resolve_urls",
  "sass",
  "slugify_urls",
  "svgo",
  "terser",
  "windi_css",
];

export function log(...lines: (string | undefined)[]) {
  console.log("----------------------------------------");
  lines.forEach((line) => line && console.log(line));
  console.log("----------------------------------------");
}

export function promptConfigUpdate(data: unknown) {
  log(
    red("deno.jsonc needs to be manually updated:"),
    dim("Use deno.json to update it automatically"),
    JSON.stringify(data, null, 2),
  );
}
