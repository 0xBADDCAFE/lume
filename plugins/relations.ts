import { merge } from "../core/utils.ts";

import type { Data, Page, Site } from "../core.ts";

export interface Options {
  /** The list of extensions this plugin applies to */
  extensions: string[];

  /** The field name used to save the page id */
  idKey: string;

  /** The field name used to save the page type */
  typeKey: string;

  /** The foreign keys per type (type => foreign_key) */
  foreignKeys: Record<string, string | [string, string]>;
}

// Default options
export const defaults: Options = {
  extensions: [".html"],
  idKey: "id",
  typeKey: "type",
  foreignKeys: {},
};

export default function (userOptions?: Partial<Options>) {
  const options = merge(defaults, userOptions);

  return (site: Site) => {
    site.preprocess(options.extensions, index);

    function index(page1: Page, pages: Page[]) {
      const data1 = page1.data;
      const [type1, foreignKey1, id1] = getForeignKey(data1);

      // Index the current page with the other pages
      pages.forEach(indexPage);

      // Index the current page with previously generated pages (if any)
      site.pages.forEach(indexPage);

      function indexPage(page2: Page) {
        if (page1 === page2) {
          return;
        }

        const data2 = page2.data;
        const [type2, foreignKey2, id2] = getForeignKey(data2);

        // Page2 has a foreign key to page1
        const directRelation = relate(
          data1,
          data2,
          foreignKey1,
          id1,
          type1,
        );

        // If it was related, do the opposite relation
        if (directRelation && type2) {
          saveMultipleRelation(
            data2,
            data1,
            type2,
          );
          return;
        }

        // Page1 has a foreign key to page2
        const reverseRelation = relate(
          data2,
          data1,
          foreignKey2,
          id2,
          type2,
        );

        // If it was related, do the opposite relation
        if (reverseRelation && type1) {
          saveMultipleRelation(
            data1,
            data2,
            type1,
          );
        }
      }

      function relate(
        rel: Data,
        data: Data,
        foreignKey?: string,
        id?: string,
        type?: string,
      ): boolean {
        if (foreignKey && type && id && data[foreignKey]) {
          const relId = data[foreignKey] as string | string[];

          // The foreign key contain an array
          if (Array.isArray(relId)) {
            if (relId.includes(id)) {
              saveMultipleRelation(rel, data, type);
              return true;
            }
            return false;
          }

          // The foreign key is a single value
          if (relId == id) {
            data[type] = rel;
            return true;
          }
        }

        return false;
      }

      function saveMultipleRelation(rel: Data, data: Data, type: string) {
        const relData = (data[type] || []) as Data[];

        if (!relData.includes(rel)) {
          relData.push(rel);
          data[type] = relData;
          // Sort by id
          relData.sort((a, b) => {
            const idA = a[options.idKey] as string;
            const idB = b[options.idKey] as string;
            return idA < idB ? -1 : 1;
          });
        }
      }
    }
  };

  function getForeignKey(data: Data): [string?, string?, string?] {
    const type = data[options.typeKey];
    if (!type) {
      return [undefined, undefined, undefined];
    }

    const foreignKey = options.foreignKeys[type];
    if (!foreignKey) {
      return [type, undefined, undefined];
    }

    if (Array.isArray(foreignKey)) {
      const [fk, id] = foreignKey;

      return [type, fk, data[id]];
    }

    return [type, foreignKey, data[options.idKey]];
  }
}
