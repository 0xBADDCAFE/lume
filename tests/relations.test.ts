import { assertSiteSnapshot, build, getSite } from "./utils.ts";
import relations from "../plugins/relations.ts";

Deno.test("relations plugin", async (t) => {
  const site = getSite({
    src: "relations",
  });

  site.use(relations({
    foreignKeys: {
      post: "post_id",
      category: ["category_id", "slug"],
      comment: "comment_id",
    },
  }));

  await build(site);
  await assertSiteSnapshot(t, site);
});
