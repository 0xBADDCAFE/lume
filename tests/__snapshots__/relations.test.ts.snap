export const snapshot = {};

snapshot[`relations plugin 1`] = `7`;

snapshot[`relations plugin 2`] = `
{
  formats: [
    {
      engines: 1,
      ext: ".tmpl.ts",
      pageLoader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".tmpl.js",
      pageLoader: [AsyncFunction: module],
    },
    {
      engines: undefined,
      ext: ".tmpl.json",
      pageLoader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: undefined,
      ext: ".json",
    },
    {
      engines: 1,
      ext: ".md",
      pageLoader: [AsyncFunction: text],
    },
    {
      componentLoader: [AsyncFunction: module],
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
    },
    {
      componentLoader: [AsyncFunction: module],
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
    },
    {
      componentLoader: [AsyncFunction: text],
      engines: 1,
      ext: ".njk",
      includesPath: "_includes",
      pageLoader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: undefined,
      ext: ".yaml",
      pageLoader: [AsyncFunction: yaml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: undefined,
      ext: ".yml",
      pageLoader: [AsyncFunction: yaml],
    },
  ],
}
`;

snapshot[`relations plugin 3`] = `
[
]
`;

snapshot[`relations plugin 4`] = `
{
  content: "<h1>Category 1</h1>

<div><p>This is the first category</p>
</div>

<h2>Posts</h2>
<ul>

</ul>
",
  data: {
    content: "This is the first category
",
    date: 1970-01-01T00:00:00.000Z,
    layout: "categories.njk",
    mergedKeys: {
      tags: "stringArray",
    },
    page: undefined,
    paginate: [Function: paginate],
    post: [
      <ref *2> {
        category: {
          content: "This is the first category
",
          date: 1970-01-01T00:00:00.000Z,
          layout: "categories.njk",
          mergedKeys: {
            tags: "stringArray",
          },
          page: undefined,
          paginate: [Function: paginate],
          post: [Circular *1],
          renderOrder: -2,
          search: Search {},
          slug: "category-1",
          tags: [
          ],
          title: "Category 1",
          type: "category",
          url: "/categories/category-1/",
        },
        category_id: "category-1",
        comment: [
          {
            content: "This is the second comment
",
            date: 1970-01-01T00:00:00.000Z,
            id: 2,
            mergedKeys: {
              tags: "stringArray",
            },
            page: undefined,
            paginate: [Function: paginate],
            post: [
              [Circular *2],
            ],
            renderOrder: -1,
            search: Search {},
            tags: [
            ],
            type: "comment",
            url: "/comments/comment-2/",
          },
        ],
        comment_id: [
          2,
        ],
        content: "Second post with category 1 and comment 2
",
        date: 1970-01-01T00:00:00.000Z,
        id: "second-post",
        layout: "posts.njk",
        mergedKeys: {
          tags: "stringArray",
        },
        page: undefined,
        paginate: [Function: paginate],
        search: Search {},
        tags: [
        ],
        type: "post",
        url: "/posts/post-2/",
      },
    ],
    renderOrder: -2,
    search: Search {},
    slug: "category-1",
    tags: [
    ],
    title: "Category 1",
    type: "category",
    url: "/categories/category-1/",
  },
  dest: {
    ext: ".html",
    path: "/categories/category-1/index",
  },
  src: {
    asset: undefined,
    ext: ".md",
    path: "/categories/category-1",
    remote: undefined,
    slug: "category-1",
  },
}
`;

snapshot[`relations plugin 5`] = `
{
  content: "<h1>Category 2</h1>

<div><p>This is the second category</p>
</div>

<h2>Posts</h2>
<ul>

</ul>
",
  data: {
    content: "This is the second category
",
    date: 1970-01-01T00:00:00.000Z,
    layout: "categories.njk",
    mergedKeys: {
      tags: "stringArray",
    },
    page: undefined,
    paginate: [Function: paginate],
    post: [
      <ref *2> {
        category: {
          content: "This is the second category
",
          date: 1970-01-01T00:00:00.000Z,
          layout: "categories.njk",
          mergedKeys: {
            tags: "stringArray",
          },
          page: undefined,
          paginate: [Function: paginate],
          post: [Circular *1],
          search: Search {},
          slug: "category-2",
          tags: [
          ],
          title: "Category 2",
          type: "category",
          url: "/categories/category-2/",
        },
        category_id: "category-2",
        comment: [
          {
            content: "This is the first comment
",
            date: 1970-01-01T00:00:00.000Z,
            id: 1,
            mergedKeys: {
              tags: "stringArray",
            },
            page: undefined,
            paginate: [Function: paginate],
            post: [
              [Circular *2],
            ],
            search: Search {},
            tags: [
            ],
            type: "comment",
            url: "/comments/comment-1/",
          },
          {
            content: "This is the third comment
",
            date: 1970-01-01T00:00:00.000Z,
            id: 3,
            mergedKeys: {
              tags: "stringArray",
            },
            page: undefined,
            paginate: [Function: paginate],
            post: [
              [Circular *2],
            ],
            search: Search {},
            tags: [
            ],
            type: "comment",
            url: "/comments/comment-3/",
          },
        ],
        comment_id: [
          1,
          3,
        ],
        content: "First post with category 2 and comments 1 and 3
",
        date: 1970-01-01T00:00:00.000Z,
        id: "first-post",
        layout: "posts.njk",
        mergedKeys: {
          tags: "stringArray",
        },
        page: undefined,
        paginate: [Function: paginate],
        renderOrder: 1,
        search: Search {},
        tags: [
        ],
        type: "post",
        url: "/posts/post-1/",
      },
    ],
    search: Search {},
    slug: "category-2",
    tags: [
    ],
    title: "Category 2",
    type: "category",
    url: "/categories/category-2/",
  },
  dest: {
    ext: ".html",
    path: "/categories/category-2/index",
  },
  src: {
    asset: undefined,
    ext: ".md",
    path: "/categories/category-2",
    remote: undefined,
    slug: "category-2",
  },
}
`;

snapshot[`relations plugin 6`] = `
{
  content: "<p>This is the first comment</p>
",
  data: {
    content: "This is the first comment
",
    date: 1970-01-01T00:00:00.000Z,
    id: 1,
    mergedKeys: {
      tags: "stringArray",
    },
    page: undefined,
    paginate: [Function: paginate],
    post: [
      <ref *1> {
        category: {
          content: "This is the second category
",
          date: 1970-01-01T00:00:00.000Z,
          layout: "categories.njk",
          mergedKeys: {
            tags: "stringArray",
          },
          page: undefined,
          paginate: [Function: paginate],
          post: [
            [Circular *1],
          ],
          search: Search {},
          slug: "category-2",
          tags: [
          ],
          title: "Category 2",
          type: "category",
          url: "/categories/category-2/",
        },
        category_id: "category-2",
        comment: [
          {
            content: "This is the first comment
",
            date: 1970-01-01T00:00:00.000Z,
            id: 1,
            mergedKeys: {
              tags: "stringArray",
            },
            page: undefined,
            paginate: [Function: paginate],
            post: [Circular *2],
            search: Search {},
            tags: [
            ],
            type: "comment",
            url: "/comments/comment-1/",
          },
          {
            content: "This is the third comment
",
            date: 1970-01-01T00:00:00.000Z,
            id: 3,
            mergedKeys: {
              tags: "stringArray",
            },
            page: undefined,
            paginate: [Function: paginate],
            post: [
              [Circular *1],
            ],
            search: Search {},
            tags: [
            ],
            type: "comment",
            url: "/comments/comment-3/",
          },
        ],
        comment_id: [
          1,
          3,
        ],
        content: "First post with category 2 and comments 1 and 3
",
        date: 1970-01-01T00:00:00.000Z,
        id: "first-post",
        layout: "posts.njk",
        mergedKeys: {
          tags: "stringArray",
        },
        page: undefined,
        paginate: [Function: paginate],
        renderOrder: 1,
        search: Search {},
        tags: [
        ],
        type: "post",
        url: "/posts/post-1/",
      },
    ],
    search: Search {},
    tags: [
    ],
    type: "comment",
    url: "/comments/comment-1/",
  },
  dest: {
    ext: ".html",
    path: "/comments/comment-1/index",
  },
  src: {
    asset: undefined,
    ext: ".md",
    path: "/comments/comment-1",
    remote: undefined,
    slug: "comment-1",
  },
}
`;

snapshot[`relations plugin 7`] = `
{
  content: "<p>This is the second comment</p>
",
  data: {
    content: "This is the second comment
",
    date: 1970-01-01T00:00:00.000Z,
    id: 2,
    mergedKeys: {
      tags: "stringArray",
    },
    page: undefined,
    paginate: [Function: paginate],
    post: [
      <ref *1> {
        category: {
          content: "This is the first category
",
          date: 1970-01-01T00:00:00.000Z,
          layout: "categories.njk",
          mergedKeys: {
            tags: "stringArray",
          },
          page: undefined,
          paginate: [Function: paginate],
          post: [
            [Circular *1],
          ],
          renderOrder: -2,
          search: Search {},
          slug: "category-1",
          tags: [
          ],
          title: "Category 1",
          type: "category",
          url: "/categories/category-1/",
        },
        category_id: "category-1",
        comment: [
          {
            content: "This is the second comment
",
            date: 1970-01-01T00:00:00.000Z,
            id: 2,
            mergedKeys: {
              tags: "stringArray",
            },
            page: undefined,
            paginate: [Function: paginate],
            post: [Circular *2],
            renderOrder: -1,
            search: Search {},
            tags: [
            ],
            type: "comment",
            url: "/comments/comment-2/",
          },
        ],
        comment_id: [
          2,
        ],
        content: "Second post with category 1 and comment 2
",
        date: 1970-01-01T00:00:00.000Z,
        id: "second-post",
        layout: "posts.njk",
        mergedKeys: {
          tags: "stringArray",
        },
        page: undefined,
        paginate: [Function: paginate],
        search: Search {},
        tags: [
        ],
        type: "post",
        url: "/posts/post-2/",
      },
    ],
    renderOrder: -1,
    search: Search {},
    tags: [
    ],
    type: "comment",
    url: "/comments/comment-2/",
  },
  dest: {
    ext: ".html",
    path: "/comments/comment-2/index",
  },
  src: {
    asset: undefined,
    ext: ".md",
    path: "/comments/comment-2",
    remote: undefined,
    slug: "comment-2",
  },
}
`;

snapshot[`relations plugin 8`] = `
{
  content: "<p>This is the third comment</p>
",
  data: {
    content: "This is the third comment
",
    date: 1970-01-01T00:00:00.000Z,
    id: 3,
    mergedKeys: {
      tags: "stringArray",
    },
    page: undefined,
    paginate: [Function: paginate],
    post: [
      <ref *1> {
        category: {
          content: "This is the second category
",
          date: 1970-01-01T00:00:00.000Z,
          layout: "categories.njk",
          mergedKeys: {
            tags: "stringArray",
          },
          page: undefined,
          paginate: [Function: paginate],
          post: [
            [Circular *1],
          ],
          search: Search {},
          slug: "category-2",
          tags: [
          ],
          title: "Category 2",
          type: "category",
          url: "/categories/category-2/",
        },
        category_id: "category-2",
        comment: [
          {
            content: "This is the first comment
",
            date: 1970-01-01T00:00:00.000Z,
            id: 1,
            mergedKeys: {
              tags: "stringArray",
            },
            page: undefined,
            paginate: [Function: paginate],
            post: [
              [Circular *1],
            ],
            search: Search {},
            tags: [
            ],
            type: "comment",
            url: "/comments/comment-1/",
          },
          {
            content: "This is the third comment
",
            date: 1970-01-01T00:00:00.000Z,
            id: 3,
            mergedKeys: {
              tags: "stringArray",
            },
            page: undefined,
            paginate: [Function: paginate],
            post: [Circular *2],
            search: Search {},
            tags: [
            ],
            type: "comment",
            url: "/comments/comment-3/",
          },
        ],
        comment_id: [
          1,
          3,
        ],
        content: "First post with category 2 and comments 1 and 3
",
        date: 1970-01-01T00:00:00.000Z,
        id: "first-post",
        layout: "posts.njk",
        mergedKeys: {
          tags: "stringArray",
        },
        page: undefined,
        paginate: [Function: paginate],
        renderOrder: 1,
        search: Search {},
        tags: [
        ],
        type: "post",
        url: "/posts/post-1/",
      },
    ],
    search: Search {},
    tags: [
    ],
    type: "comment",
    url: "/comments/comment-3/",
  },
  dest: {
    ext: ".html",
    path: "/comments/comment-3/index",
  },
  src: {
    asset: undefined,
    ext: ".md",
    path: "/comments/comment-3",
    remote: undefined,
    slug: "comment-3",
  },
}
`;

snapshot[`relations plugin 9`] = `
{
  content: '<div><p>First post with category 2 and comments 1 and 3</p>
</div>

  
<h2>Category:</h2>
<a href="/categories/category-2/">Category 2</a>
<p>This is the second category</p>


<ul>

  <li><p>This is the first comment</p></li>

  <li><p>This is the third comment</p></li>

</ul>',
  data: {
    category: <ref *1> {
      content: "This is the second category
",
      date: 1970-01-01T00:00:00.000Z,
      layout: "categories.njk",
      mergedKeys: {
        tags: "stringArray",
      },
      page: undefined,
      paginate: [Function: paginate],
      post: [
        <ref *2> {
          category: [Circular *1],
          category_id: "category-2",
          comment: [
            {
              content: "This is the first comment
",
              date: 1970-01-01T00:00:00.000Z,
              id: 1,
              mergedKeys: {
                tags: "stringArray",
              },
              page: undefined,
              paginate: [Function: paginate],
              post: [
                [Circular *2],
              ],
              search: Search {},
              tags: [
              ],
              type: "comment",
              url: "/comments/comment-1/",
            },
            {
              content: "This is the third comment
",
              date: 1970-01-01T00:00:00.000Z,
              id: 3,
              mergedKeys: {
                tags: "stringArray",
              },
              page: undefined,
              paginate: [Function: paginate],
              post: [
                [Circular *2],
              ],
              search: Search {},
              tags: [
              ],
              type: "comment",
              url: "/comments/comment-3/",
            },
          ],
          comment_id: [
            1,
            3,
          ],
          content: "First post with category 2 and comments 1 and 3
",
          date: 1970-01-01T00:00:00.000Z,
          id: "first-post",
          layout: "posts.njk",
          mergedKeys: {
            tags: "stringArray",
          },
          page: undefined,
          paginate: [Function: paginate],
          renderOrder: 1,
          search: Search {},
          tags: [
          ],
          type: "post",
          url: "/posts/post-1/",
        },
      ],
      search: Search {},
      slug: "category-2",
      tags: [
      ],
      title: "Category 2",
      type: "category",
      url: "/categories/category-2/",
    },
    category_id: "category-2",
    comment: [
      {
        content: "This is the first comment
",
        date: 1970-01-01T00:00:00.000Z,
        id: 1,
        mergedKeys: {
          tags: "stringArray",
        },
        page: undefined,
        paginate: [Function: paginate],
        post: [
          <ref *2> {
            category: <ref *1> {
              content: "This is the second category
",
              date: 1970-01-01T00:00:00.000Z,
              layout: "categories.njk",
              mergedKeys: {
                tags: "stringArray",
              },
              page: undefined,
              paginate: [Function: paginate],
              post: [
                [Circular *2],
              ],
              search: Search {},
              slug: "category-2",
              tags: [
              ],
              title: "Category 2",
              type: "category",
              url: "/categories/category-2/",
            },
            category_id: "category-2",
            comment: [Circular *3],
            comment_id: [
              1,
              3,
            ],
            content: "First post with category 2 and comments 1 and 3
",
            date: 1970-01-01T00:00:00.000Z,
            id: "first-post",
            layout: "posts.njk",
            mergedKeys: {
              tags: "stringArray",
            },
            page: undefined,
            paginate: [Function: paginate],
            renderOrder: 1,
            search: Search {},
            tags: [
            ],
            type: "post",
            url: "/posts/post-1/",
          },
        ],
        search: Search {},
        tags: [
        ],
        type: "comment",
        url: "/comments/comment-1/",
      },
      {
        content: "This is the third comment
",
        date: 1970-01-01T00:00:00.000Z,
        id: 3,
        mergedKeys: {
          tags: "stringArray",
        },
        page: undefined,
        paginate: [Function: paginate],
        post: [
          <ref *2> {
            category: <ref *1> {
              content: "This is the second category
",
              date: 1970-01-01T00:00:00.000Z,
              layout: "categories.njk",
              mergedKeys: {
                tags: "stringArray",
              },
              page: undefined,
              paginate: [Function: paginate],
              post: [
                [Circular *2],
              ],
              search: Search {},
              slug: "category-2",
              tags: [
              ],
              title: "Category 2",
              type: "category",
              url: "/categories/category-2/",
            },
            category_id: "category-2",
            comment: [Circular *3],
            comment_id: [
              1,
              3,
            ],
            content: "First post with category 2 and comments 1 and 3
",
            date: 1970-01-01T00:00:00.000Z,
            id: "first-post",
            layout: "posts.njk",
            mergedKeys: {
              tags: "stringArray",
            },
            page: undefined,
            paginate: [Function: paginate],
            renderOrder: 1,
            search: Search {},
            tags: [
            ],
            type: "post",
            url: "/posts/post-1/",
          },
        ],
        search: Search {},
        tags: [
        ],
        type: "comment",
        url: "/comments/comment-3/",
      },
    ],
    comment_id: [
      1,
      3,
    ],
    content: "First post with category 2 and comments 1 and 3
",
    date: 1970-01-01T00:00:00.000Z,
    id: "first-post",
    layout: "posts.njk",
    mergedKeys: {
      tags: "stringArray",
    },
    page: undefined,
    paginate: [Function: paginate],
    renderOrder: 1,
    search: Search {},
    tags: [
    ],
    type: "post",
    url: "/posts/post-1/",
  },
  dest: {
    ext: ".html",
    path: "/posts/post-1/index",
  },
  src: {
    asset: undefined,
    ext: ".md",
    path: "/posts/post-1",
    remote: undefined,
    slug: "post-1",
  },
}
`;

snapshot[`relations plugin 10`] = `
{
  content: '<div><p>Second post with category 1 and comment 2</p>
</div>

  
<h2>Category:</h2>
<a href="/categories/category-1/">Category 1</a>
<p>This is the first category</p>


<ul>

  <li><p>This is the second comment</p></li>

</ul>',
  data: {
    category: <ref *1> {
      content: "This is the first category
",
      date: 1970-01-01T00:00:00.000Z,
      layout: "categories.njk",
      mergedKeys: {
        tags: "stringArray",
      },
      page: undefined,
      paginate: [Function: paginate],
      post: [
        <ref *2> {
          category: [Circular *1],
          category_id: "category-1",
          comment: [
            {
              content: "This is the second comment
",
              date: 1970-01-01T00:00:00.000Z,
              id: 2,
              mergedKeys: {
                tags: "stringArray",
              },
              page: undefined,
              paginate: [Function: paginate],
              post: [
                [Circular *2],
              ],
              renderOrder: -1,
              search: Search {},
              tags: [
              ],
              type: "comment",
              url: "/comments/comment-2/",
            },
          ],
          comment_id: [
            2,
          ],
          content: "Second post with category 1 and comment 2
",
          date: 1970-01-01T00:00:00.000Z,
          id: "second-post",
          layout: "posts.njk",
          mergedKeys: {
            tags: "stringArray",
          },
          page: undefined,
          paginate: [Function: paginate],
          search: Search {},
          tags: [
          ],
          type: "post",
          url: "/posts/post-2/",
        },
      ],
      renderOrder: -2,
      search: Search {},
      slug: "category-1",
      tags: [
      ],
      title: "Category 1",
      type: "category",
      url: "/categories/category-1/",
    },
    category_id: "category-1",
    comment: [
      {
        content: "This is the second comment
",
        date: 1970-01-01T00:00:00.000Z,
        id: 2,
        mergedKeys: {
          tags: "stringArray",
        },
        page: undefined,
        paginate: [Function: paginate],
        post: [
          <ref *2> {
            category: <ref *1> {
              content: "This is the first category
",
              date: 1970-01-01T00:00:00.000Z,
              layout: "categories.njk",
              mergedKeys: {
                tags: "stringArray",
              },
              page: undefined,
              paginate: [Function: paginate],
              post: [
                [Circular *2],
              ],
              renderOrder: -2,
              search: Search {},
              slug: "category-1",
              tags: [
              ],
              title: "Category 1",
              type: "category",
              url: "/categories/category-1/",
            },
            category_id: "category-1",
            comment: [Circular *3],
            comment_id: [
              2,
            ],
            content: "Second post with category 1 and comment 2
",
            date: 1970-01-01T00:00:00.000Z,
            id: "second-post",
            layout: "posts.njk",
            mergedKeys: {
              tags: "stringArray",
            },
            page: undefined,
            paginate: [Function: paginate],
            search: Search {},
            tags: [
            ],
            type: "post",
            url: "/posts/post-2/",
          },
        ],
        renderOrder: -1,
        search: Search {},
        tags: [
        ],
        type: "comment",
        url: "/comments/comment-2/",
      },
    ],
    comment_id: [
      2,
    ],
    content: "Second post with category 1 and comment 2
",
    date: 1970-01-01T00:00:00.000Z,
    id: "second-post",
    layout: "posts.njk",
    mergedKeys: {
      tags: "stringArray",
    },
    page: undefined,
    paginate: [Function: paginate],
    search: Search {},
    tags: [
    ],
    type: "post",
    url: "/posts/post-2/",
  },
  dest: {
    ext: ".html",
    path: "/posts/post-2/index",
  },
  src: {
    asset: undefined,
    ext: ".md",
    path: "/posts/post-2",
    remote: undefined,
    slug: "post-2",
  },
}
`;
